cli/util/ssh.go

```diff
@@ -0,0 +1,183 @@
+
+	if remotePort := strings.Split(host, ":"); len(remotePort) == 2 {
+		port = remotePort[1]
+		userAtHost = remotePort[0]
+	} else {
+		userAtHost = host
+	}
+
+	args := c.StringSlice("ssh-args")
+	// Check if port is specified explicitly with the --ssh-args flag
+	addPort := true
+	for _, arg := range args {
+		if strings.Contains(arg, "-p") {
+			addPort = false
+			break
+		}
+	}
+	if addPort {
+		args = append(args, "-p", port)
+	}
+	args = append(args, userAtHost)
+	args = append(
+		args,
+		"/bin/sh",
+		"-c",
+		command)
+
+	cmd := exec.Command("ssh", args...)
+
+	// Simply connect stdin/stderr
+	cmd.Stdin = os.Stdin
+	cmd.Stderr = os.Stderr
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, errors.New("Error redirecting stdout on exec")
+	}
+
+	// Disable tty echo before starting
+	term, err := DisableEcho(int(os.Stdin.Fd()))
+	if err == nil {
+		sigChan = make(chan os.Signal, 1)
+		errChan = make(chan error, 1)
+		// Make sure that echo is enabled if the process gets
+		// interrupted
+		signal.Notify(sigChan)
+		go EchoSigHandler(_ctx, sigChan, errChan, term)
+	} else if err != syscall.ENOTTY {
+		return nil, err
+	}
+
+	if err := cmd.Start(); err != nil {
+		return nil, err
+	}
+
+	// Wait for 60 seconds for ssh to establish connection
+	err = waitForBufferSignal(stdout, os.Stdout, sshConnectedToken, 2*time.Minute)
+	if err != nil {
+		_ = cmd.Process.Kill()
+		return nil, errors.Wrap(err,
+			"Error waiting for ssh session to be established.")
+	}
+	return &SSHCommand{
+		ctx:     _ctx,
+		Cmd:     cmd,
+		Stdout:  stdout,
+		cancel:  cancel,
+		sigChan: sigChan,
+		errChan: errChan,
+	}, nil
+}
+
+func (s *SSHCommand) EndSSHCommand() error {
+	if s.Cmd.ProcessState != nil && s.Cmd.ProcessState.Exited() {
+		return errors.New("SSH session closed unexpectedly")
+	}
+
+	if err := s.Cmd.Wait(); err != nil {
+		return errors.Wrap(err,
+			"SSH session closed with error")
+	}
+
+	if s.sigChan != nil {
+		signal.Stop(s.sigChan)
+		s.cancel()
+		if err := <-s.errChan; err != nil {
+			return err
+		}
+	} else {
+		s.cancel()
+	}
+
+	return nil
+}
+
+// Reads from src waiting for the string specified by signal, writing all other
+// output appearing at src to sink. The function returns an error if occurs
+// reading from the stream or the deadline exceeds.
+func waitForBufferSignal(src io.Reader, sink io.Writer,
+	signal string, deadline time.Duration) error {
+
+	var err error
+	errChan := make(chan error)
+
+	go func() {
+		stdoutRdr := bufio.NewReader(src)
+		for {
+			line, err := stdoutRdr.ReadString('\n')
+			if err != nil {
+				errChan <- err
+				break
+			}
+			if strings.Contains(line, signal) {
+				errChan <- nil
+				break
+			}
+			_, err = sink.Write([]byte(line + "\n"))
+			if err != nil {
+				errChan <- err
+				break
+			}
+		}
+	}()
+
+	select {
+	case err = <-errChan:
+		// Error from goroutine
+	case <-time.After(deadline):
+		err = errors.New("Input deadline exceeded")
```

Comment:
We could also reuse the deadline from the context in the parent scope here instead of a fixed duration.
```suggestion
	case <-ctx.Done():
		err = ctx.Err()
```

Reply:
Sorry for being unclear. My point was to not have multiple sources of deadlines. The context itself can be used to provide a deadline (see [context.WithTimeout](https://pkg.go.dev/context#WithTimeout)). In the above select case we allocate a new timer and ignore the context which is used in parallel by other go routines.